scripts/changesummary.py

```diff
@@ -0,0 +1,540 @@
+
+        # Split the Key into 2 columns for `Parent` and `Child` in order
+        # to group keys with the same parents together to summarize the changes
+        # by parent.
+        parent_child_df = combined_docs["Key"].str.rsplit(".", 1, expand=True)
+        # Rename the columns and join them with the combined_docs dataframe.
+        # If we only have a `Parent` column, it means that the Key doesn't have
+        # any children.
+        if len(parent_child_df.columns) == 1:
+            parent_child_df.columns = ["Parent"]
+        else:
+            parent_child_df.columns = ["Parent", "Child"]
+        combined_docs = combined_docs.join(parent_child_df)
+
+        # Create a new column `Added` to identify rows which have new keys.
+        combined_docs["Added"] = np.where(
+            combined_docs["CurrentValue"].isnull(), True, False
+        )
+
+        # Create a new column `Deleted` to identify rows which have deleted keys.
+        combined_docs["Deleted"] = np.where(
+            combined_docs["NewValue"].isnull(), True, False
+        )
+
+        # Aggregate the keys added by grouping keys with the same parents
+        # together to summarize the changes by parent rather than by key.
+        parent_added_agg = (
+            combined_docs.groupby("Parent")
+            .Added.value_counts(normalize=True)
+            .reset_index(name="Percent")
+        )
+
+        # Add a column NumLevels to inicate the number of levels in the tree
+        # which will allow us to sort the parents in hierarchical order.
+        parent_added_agg["NumLevels"] = (
+            parent_added_agg["Parent"].str.split(".").apply(lambda x: len(x))
+        )
+
+        # Aggregate the keys deleted by grouping keys with the same parents
+        # together to summarize the changes by parent rather than by key.
+        parent_deleted_agg = (
+            combined_docs.groupby("Parent")
+            .Deleted.value_counts(normalize=True)
+            .reset_index(name="Percent")
+        )
+
+        # Add a column NumLevels to inicate the number of levels in the tree
+        # which will allow us to sort the parents in hierarchical order.
+        parent_deleted_agg["NumLevels"] = (
+            parent_added_agg["Parent"].str.split(".").apply(lambda x: len(x))
+        )
+
+        # Create a list of all parents that have been added in hierarchical
+        # order. When `Percent` is 1, it means that the parent is new as all
+        # children keys have been added.
+        all_added = (
+            parent_added_agg[
+                (parent_added_agg["Percent"] == 1) & (parent_added_agg["Added"] == True)
+            ][["Parent", "NumLevels"]]
+            .sort_values("NumLevels", ascending=True)
+            .Parent.to_list()
+        )
+
+        # Create a list of all parents that have been deleted in hierarchical
+        # order. When `Percent` is 1, it means that the parent is new as all
+        # children keys have been added.
```

Comment:
I see this comment is the same as above, but shouldn't it be "completely deleted" or something like that?

Reply:
Good catch! Fixed.